File: flutter_app/lib/models/chat_message.dart
class ChatMessage {
  ChatMessage({
    required this.id,
    required this.conversationId,
    required this.senderId,
    required this.content,
    required this.createdAt,
    this.messageType = 'text',
    this.isRead = false,
  });

  final int id;
  final int conversationId;
  final int senderId;
  final String content;
  final String createdAt;
  final String messageType;
  final bool isRead;

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    final rawRead = json['is_read'];
    return ChatMessage(
      id: _asInt(json['id']),
      conversationId: _asInt(json['conversation_id']),
      senderId: _asInt(json['sender_id']),
      content: (json['content'] ?? '') as String,
      createdAt: (json['created_at'] ?? '') as String,
      messageType: (json['message_type'] ?? 'text') as String,
      isRead: rawRead is bool ? rawRead : _asInt(rawRead) == 1,
    );
  }
}

int _asInt(dynamic value) {
  if (value is int) {
    return value;
  }
  return int.tryParse('$value') ?? 0;
}

File: frontend/src/models/chatMessage.ts
import {
  asBoolean,
  asInt,
  asRecord,
  asString,
} from './parse';

export interface ChatMessage {
  id: number;
  conversationId: number;
  senderId: number;
  content: string;
  createdAt: string;
  messageType: string;
  isRead: boolean;
}

export function parseChatMessage(value: unknown): ChatMessage {
  const json = asRecord(value);
  return {
    id: asInt(json.id),
    conversationId: asInt(json.conversation_id),
    senderId: asInt(json.sender_id),
    content: asString(json.content),
    createdAt: asString(json.created_at),
    messageType: asString(json.message_type, 'text'),
    isRead: asBoolean(json.is_read),
  };
}
